### ISSUE TITLE
Calling map on maps many times causes segfault

### ISSUE THREAD

**Issue by AFK-debug-9** [python/cpython#103503]
# Crash report
When the following code is ran, It causes a Segmentation Fault:
```py
a = []
for x in range(100_000):
    a = map(int, a)
list(a)
```
Now the `list` call calls `Python/bltinmodule.c:map_next`.
If we run the code with gdb, we get the following: 
```
Program received signal SIGSEGV, Segmentation fault.
```
But we also get: 
```
0x00005555557bbd6e in map_next (lz=0x7ffff7099ba0) at Python/bltinmodule.c:1342
```
To inspect the call stack, we can type `where` into gdb and we get loads of function calls, all seeming to be coming from line 1363, which is:
```c
PyObject *val = Py_TYPE(it)->tp_iternext(it);
```
And the value of `Py_TYPE(it)` is a `map` and it calls `map_next` which causes the recursion 

# Error message
Segmentation Fault

# Core file
Core file: [Google Drive Link](https://drive.google.com/file/d/1nuMfIq_l2zgdwJKbYnbRqUByjf48cuNG/view?usp=sharing)

# Possible fix
We could add a recursion limit like we have with regular python functions.

# Your environment
- CPython versions tested on: 3.9, 3.10, 3.12.0a5+
- Operating system and architecture: Windows & Linux, x86-64

**Comment by sunmy2019** [comment]
Looks like a stack overflow.

On my computer, the final call stack is number 65510, very close to 65536, and it seems that one stack frame uses 64 Bytes, which is exactly ~4MB .

> Looks like a stack overflow.

This can be verified by increasing the stack size. It works on my computer when I increase the stack size to 100MB.
https://stackoverflow.com/questions/15887787/increase-stack-size-for-a-program-to-run

**Comment by sunmy2019** [comment]
Calling `__next__` is handled in C. Thus only the C stack increases, not the Python stack. I don't think any checks were performed.

**Comment by AFK-debug-9** [comment]
Yes this is evident as the call stack is near 65k, as it crashed showing there are no checks involved, we could add a check if it goes over 65k calls and raise a `RecursionError` like:
```
Traceback (most recent call last):
  File "<main.py>", line 1, in <module>
  File "<builtin>", in map
  File "<builtin>", in map
  File "<builtin>", in map
  [Previous line repeated 65000 more times]
RecursionError: maximum recursion depth exceeded
```

**Comment by AFK-debug-9** [comment]
But it has not been solved

**Comment by pochmann** [comment]
Does it need to be solved? Is it a problem? What real code does that? Such deep nesting is highly inefficient, you wouldn't want to do that anyway.

**Comment by terryjreedy** [comment]
@rhettinger This seems similar to the [nested itertools issue](https://github.com/python/cpython/issues/58218) that you closed (as won't fix).

**Comment by rhettinger** [comment]
@pochmann As you noted, this doesn't seem to arise in practice, so I'm don't worry about it.  

If someone really cares about this, this best solution would be for the interpreter to monitor the actual stack usage and respond appropriately when running low on space.  This is similar to what people do to monitor free disk space on servers.  Implementing this would be real challenge though as it requires substantial expertise with compilers and operating systems.  It is in the doable but hard category.

The other challenge would be to define what we mean by "responding appropriately".  To get the system back to a usable state, the stack would have to be cleared-off without breaking invariants.  The only way I know to do this is by snapshotting.  That would be a somewhat heavy handed solution to a problem that almost no user ever has.

**Comment by carljm** [comment]
If anyone wants to pursue stack overflow prevention, it would be a good idea to start with understanding the [rejected PEP 651](https://peps.python.org/pep-0651/) and the [reasons for its rejection](https://mail.python.org/archives/list/python-dev@python.org/thread/75BFSBM5AJWXOF5OSPLMJQSTP3TDOKRP/). (Note that one reason for rejection was concern about multiple Python call frames in a single C frame, which we already now have.)

**Comment by AFK-debug-9** [comment]
We could do some crude pointer arithmetic

**Comment by picnixz** [comment]
I'm going to close this one as the others were closed. Preventing those cases is probably not worth the cost of slowing down regular cases. Unless a concrete real-world example is impeded by this, similar issues should be closed as well.

**Comment by markshannon** [comment]
We now have fast (a few machine instructions) stack overflow checking available, so let's fix this.
